websites/R/Rate Your Music/presence.ts

```diff
@@ -0,0 +1,209 @@
+			"#page_genre_section_name h1"
+		);
+		if (genreNameElement) {
+			const genreName = genreNameElement.textContent?.trim();
+			if (genreName) {
+				presenceData.details = "Viewing Genre:";
+				presenceData.state = genreName;
+			}
+		}
+	} else if (pathname.startsWith("/artist/")) {
+		const artistNameElement = document.querySelector("h1.artist_name_hdr");
+		if (artistNameElement) {
+			const artistName = artistNameElement.textContent?.trim();
+			if (artistName) {
+				presenceData.details = "Viewing Artist:";
+				presenceData.state = artistName;
+			}
+		}
+	} else if (pathname.startsWith("/~")) {
+		const usernameElement = document.querySelector("#profilename");
+		if (usernameElement) {
+			const username = usernameElement.textContent?.trim();
+			if (username) {
+				presenceData.details = "Viewing User:";
+				presenceData.state = username;
+			}
+		}
+	} else if (pathname.startsWith("/list/")) {
+		const listTitleElement = document.querySelector(
+				"h1[style='font-size:2.1em']"
+			),
+			listOwner = pathname.split("/")[2];
+		if (listTitleElement) {
+			const listTitle = listTitleElement.textContent?.trim();
+			if (listTitle && listOwner) {
+				presenceData.details = "Viewing List:";
+				presenceData.state = `${listTitle} by ${listOwner}`;
+			}
+		}
+	} else if (pathname.startsWith("/rdescriptor/set")) {
+		const albumElement = document.querySelector("a.album"),
+			artistElement = document.querySelector("a.artist");
+		if (albumElement && artistElement) {
+			presenceData.details = "Voting on Descriptors";
+			presenceData.state = `${albumElement.textContent?.trim()} by ${artistElement.textContent?.trim()}`;
+		}
+	} else if (
+		pathname.startsWith("/lists/edit") &&
+		document.querySelector("span")
+	)
+		presenceData.details = "Editing List:";
+	else if (pathname.startsWith("/search")) {
+		presenceData.details = "Searching for:";
+		presenceData.state = new URLSearchParams(pathname).get("searchterm");
+	} else if (pathname.startsWith("/release/")) {
+		const albumTitleElement = document.querySelector(".album_title");
+		if (
+			albumTitleElement &&
+			document.querySelector(".album_artist_small a.artist")
+		) {
+			presenceData.details = "Viewing Release:";
+			presenceData.state = `${albumTitleElement.textContent?.trim()}`;
+		}
+	} else if (pathname.startsWith("/collection/")) {
+		presenceData.details = "Viewing Collection:";
+		presenceData.state = `${pathname.split("/")[2]}`;
+	} else if (pathname.startsWith("/find_similar_users")) {
+		presenceData.details = "Viewing Compatibility:";
+		presenceData.state = new URLSearchParams(pathname).get("user");
+	}
+
+	for (const [path, data] of Object.entries(pages)) {
+		if (pathname === path) {
+			presenceData = { ...presenceData, ...data };
+			break;
+		}
+	}
+
+	new Presence({
+		clientId: "1213365863470866462",
+	}).setActivity(presenceData);
+}
+
+updatePresence();
+
+document.addEventListener("UpdateData", updatePresence);
```

Comment:
This still has the same problems. `new Presence` may cause performance problems over time when done multiple times.
```ts
const presence = new Presence(...);

presence.on("UpdateData", updatePresences);

function updatePresences() {
  // ...
  presence.setActivity(presenceData)
}
```